
############################## Load libraries #######################################
library(getopt)
library(optparse)
library(parallel)
library(tidyverse)
library(ggplot2)
library(dplyr)
library(GenomicFeatures)
library(HiCDCPlus)
library(gridExtra)
library(grid)
library(VennDiagram)

library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")

############################## Set up s"cript options #######################################
# Read in command line opts
option_list <- list(
  make_option(c("-r", "--runtype"), action = "store", type = "character", help = "Specify whether running through through 'nextflow' in order to switch paths"),
  make_option(c("-c", "--cores"), action = "store", type = "integer", help = "Number of CPUs"),
  make_option(c("", "--verbose"), action = "store", type = "logical", help = "Verbose", default = FALSE)
)

opt_parser = OptionParser(option_list = option_list)
opt <- parse_args(opt_parser)
if(opt$verbose) print(opt)


# Set paths and load data
{
  if(length(commandArgs(trailingOnly = TRUE)) == 0){
    cat('No command line arguments provided, paths are set for running interactively in Rstudio server\n')
    
    ncores = 8
    
    # for testing with one sample
    plot_path = "./output/NF-hichip-downstream/NF_HiChip_r1/HicDCPlus_output_investigating/plots/"
    rds_path = "./output/NF-hichip-downstream/NF_HiChip_r1/HicDCPlus_output_investigating/rds_files/"
    data_path = "./output/NF-hichip-downstream/NF_HiChip_r1/HicDCPlus_output/" # for HiCDCPlus output of sample 1
    
    # for testing with stringent consensus file
    data_path = "./output/NF-hichip-downstream/AllSamples/HicDCPlus_diff_interactions/rds_files/" # for consensus interactions across 3 NF samples - all 3
    plot_path = "./output/NF-hichip-downstream/AllSamples/HicDCPlus_output_investigating/stringent/plots/"
    rds_path = "./output/NF-hichip-downstream/AllSamples/HicDCPlus_output_investigating/stringent/rds_files/"
    
    # for testing with less stringent consensus file
    data_path = "./output/NF-hichip-downstream/AllSamples/HicDCPlus_diff_interactions/rds_files/" # for consensus interactions across 3 NF samples - 2 of 3
    plot_path = "./output/NF-hichip-downstream/AllSamples/HicDCPlus_output_investigating/less_stringent/plots/"
    rds_path = "./output/NF-hichip-downstream/AllSamples/HicDCPlus_output_investigating/less_stringent/rds_files/"
    
    # for the other data inputs
    data_path = "./output/NF-hichip-downstream/bins/peaks_intersect/" # for filtering by peaks
    data_path = "./output/NF-hichip-downstream/bins/promoters_intersect/" # for filtering by promoters
    data_path = "./output/NF-hichip-downstream/bins/bed_files/" # for bins
    
  } else if (opt$runtype == "nextflow"){
    cat('pipeline running through Nextflow\n')
    
    plot_path = "./plots/"
    rds_path = "./rds_files/"
    data_path = "./input/"
    ncores = opt$cores
    
  } else {
    stop("--runtype must be set to 'nextflow'")
  }
  
  cat(paste0("script ran with ", ncores, " cores\n")) 
  dir.create(plot_path, recursive = T)
  dir.create(rds_path, recursive = T)
}


######################################   Read in data   ####################################################

print("reading in data...")

# read in all bins
print("All bins:")
bins <- data.table::fread(paste0(data_path, "bins.bed"))
colnames(bins) <- c("chr", "start", "end", "bin_ID")
head(bins)

# read in peaks intersected with bins
print("Peaks intersected with bins:")
peaks_bins <- data.table::fread(paste0(data_path, "FullData_PeakSet_bins_intersected.bed"))
colnames(peaks_bins) <- c("bin_chr", "bin_start", "bin_end", "bin_ID", "peak_chr", "peak_start", "peak_end", "peak_ID", "dunno", "strand")
head(peaks_bins)
dim(peaks_bins)

# read in promoters intersected with bins
print("Promoters intersected with bins:")
promoters_bins <- data.table::fread(paste0(data_path, "tag_chroms_bins_intersected.bed"))
colnames(promoters_bins) <- c("bin_chr", "bin_start", "bin_end", "bin_ID", "promoter_chr", "promoter_start", "promoter_end", "dunno", "strand", "gene_ID", "gene_name")
head(promoters_bins)
dim(promoters_bins)

# read in HiCDCPlus output
print("HiCDCPlus output:")
# interactions <- data.table::fread(paste0(data_path, "rds_files/NF_HiChip_r1_HiCDC_output_filtered.txt"))
interactions <- data.table::fread(paste0(data_path, "Consensus_interactions.txt"))
#interactions <- data.table::fread(paste0(data_path, "Consensus_interactions_stringent.txt"))
interactions <- interactions[,-1]
head(interactions)
dim(interactions)

print("data read in!")

##############  Looking at anchors and how many times they appear    ###########################################

print("Investigating interaction anchor frequencies...")

# how many significant interactions are there
how_many_interactions <- nrow(interactions)
print(paste0("Number of sig interactions: ", how_many_interactions))

# extract all anchors in the significant interactions
anchors_I <- interactions[, 1:3]
colnames(anchors_I) <- c("chr", "start", "end")
anchors_I <- anchors_I %>% mutate(bin_ID = paste0(chr, "-", start+1, "-", end))
anchors_J <- interactions[, 4:6]
colnames(anchors_J) <- c("chr", "start", "end")
anchors_J <- anchors_J %>% mutate(bin_ID = paste0(chr, "-", start+1, "-", end))

interactions <- interactions %>% 
  mutate(anchor_I = anchors_I$bin_ID) %>%
  mutate(anchor_J = anchors_J$bin_ID)

# rbind all these anchors together
all_anchors <- rbind(anchors_I, anchors_J)
print("All anchors:")
print(head(all_anchors))

# frequency at which these anchors appear
print("Frequency at which anchors occur in significant interactions: ")
print(table(table(all_anchors$bin_ID)))

png(paste0(plot_path, "freq_of_anchors_in_interactions.png"), width=60, height=30, units = 'cm', res = 200)
plot(table(table(all_anchors$bin_ID)))
graphics.off()

# pull out the anchors that appear more than 100 times - can vary this threshold
highly_interacting_anchors <- table(all_anchors$bin_ID)[table(all_anchors$bin_ID) > 100]
print("highly interacting anchors: ")
print(highly_interacting_anchors)

##############  Overlap of anchors with any promoter and any peak called across all scATAC data    ###########################################

## how many bins in each type of data
bins_numbers <- data.frame(
  Total_Bins = length(unique(bins$bin_ID)),
  Interacting_Bins = length(unique(all_anchors$bin_ID)),
  Promoter_Bins = length(unique(promoters_bins$bin_ID)),
  Peak_Bins = length(unique(peaks_bins$bin_ID))
)

png(paste0(plot_path, 'bin_counts_table.png'), height = 10, width = 20, units = 'cm', res = 400)
grid.arrange(tableGrob(bins_numbers, rows=NULL, theme = ttheme_minimal()))
graphics.off()

# sanity check that all anchors are in the bins
if (sum(unique(all_anchors$bin_ID) %in% bins$bin_ID)){
  print("All anchors are in the bins!")
}else{stop("PROBLEM: NOT ALL ANCHORS ARE IN THE BINS!")}

# Venn diagram of bins between interactions, peaks and genes
venn.diagram(
  x = list(unique(all_anchors$bin_ID), unique(promoters_bins$bin_ID), unique(peaks_bins$bin_ID)),
  category.names = c("Interactions", "Promoters", "Peaks"),
  filename = paste0(plot_path, 'bins_venn_diagram.png'),
  output=TRUE, disable.logging = TRUE,
  # Output features
  imagetype="png",
  height = 600, 
  width = 600, 
  resolution = 300,
  compression = "lzw",
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  # Numbers
  cex = .6,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)
graphics.off()



##############  FILTER INTERACTIONS    ###########################################

## 1) Remove interactions in which the bins involved are NOT annotated as peaks or promoters

# all annotated bins
annotated_bins <- unique(c(peaks_bins$bin_ID, promoters_bins$bin_ID))
length(annotated_bins)

# filter interactions so only include ones where both bins are annotated
filtered_interactions <- interactions %>% filter(anchor_I %in% annotated_bins & anchor_J %in% annotated_bins)
dim(filtered_interactions)
dim(interactions)

## 2) Include only interactions where there is at least ONE promoter and ONE peak on either side of each interactions

# split interactions to 2 dataframes: one where both bins are associated with peaks + promoters, and one where at least one bin is associated to only peak or only promoters
filtered_interactions <- filtered_interactions %>% rownames_to_column('ID')
bins_in_both <- filtered_interactions %>% filter(anchor_I %in% peaks_bins$bin_ID & anchor_I %in% promoters_bins$bin_ID | anchor_J %in% peaks_bins$bin_ID & anchor_J %in% promoters_bins$bin_ID)
bins_in_one <- filtered_interactions[!filtered_interactions$ID %in% bins_in_both$ID]

# in the case where at least one bin is only associated to one, make sure one anochor is associated to peak and one to promoter
filtered_bins_in_one <- bins_in_one[bins_in_one[['anchor_I']] %in% peaks_bins$bin_ID & bins_in_one[['anchor_J']] %in% promoters_bins$bin_ID | 
  bins_in_one[['anchor_J']] %in% peaks_bins$bin_ID & bins_in_one[['anchor_I']] %in% promoters_bins$bin_ID,]

# rbind this filtered df with the bins_in_both df
filtered_interactions <- rbind(filtered_bins_in_one, bins_in_both)

nrow(filtered_interactions)

##############  Read in genes of interest    ###########################################
# from Alex's GMs analysis, GMs made on ss8 data, these are the ones used for coexpression and temporal analysis

# PPR specific GMs
# GM12; AKR1D1, ATP1A1, ATP1B1, ATP2B1, B3GNT7, CCDC25, CGNL1, DAG1, EMILIN2, ENSGALG00000004814, EPCAM, FAM184B, FAM89A, GATA2, GATA3, IRAK2, IRF6, MAP7, METRNL, MPZL3, NET1, PLEKHA5, POGLUT2, PPFIBP1, RGN, SLC16A10, SLC25A4, TSPAN13, TTC39A, UNC5B, Z-TJP2
# GM14; BRINP1, CITED4, DLX5, DLX6, ENSGALG00000023936, ENSGALG00000040010, FN1, HESX1, HIF1A, KCNAB1, LAMB1, LRP11, NFKB1, PAX6, PITX1, PITX2, SEM1, SFRP1, SH3D19, SHISA2, SIX3, SPON1, SST, WDR1, Z-HAPLN1
# GM13; AKAP12, ASS1, BASP1, CD99, ENSGALG00000011296, ENSGALG00000041054, ENSGALG00000042443, EYA2, FERMT2, LGMN, METTL24, NR2F2, NUCKS1, SIX1


PPR_genes <- c("AKR1D1", "ATP1A1", "ATP1B1", "ATP2B1", "B3GNT7", "CCDC25", "CGNL1", "DAG1", "EMILIN2", "ENSGALG00000004814", "EPCAM", "FAM184B", "FAM89A", "GATA2", "GATA3", "IRAK2", "IRF6", "MAP7", "METRNL", "MPZL3", "NET1", "PLEKHA5", "POGLUT2", "PPFIBP1", "RGN", "SLC16A10", "SLC25A4", "TSPAN13", "TTC39A", "UNC5B", "Z-TJP2",
               "BRINP1", "CITED4", "DLX5", "DLX6", "ENSGALG00000023936", "ENSGALG00000040010", "FN1", "HESX1", "HIF1A", "KCNAB1", "LAMB1", "LRP11", "NFKB1", "PAX6", "PITX1", "PITX2", "SEM1", "SFRP1", "SH3D19", "SHISA2", "SIX3", "SPON1", "SST", "WDR1", "Z-HAPLN1",
               "AKAP12", "ASS1", "BASP1", "CD99", "ENSGALG00000011296", "ENSGALG00000041054", "ENSGALG00000042443", "EYA2", "FERMT2", "LGMN", "METTL24", "NR2F2", "NUCKS1", "SIX1")

# NC specific GMs
# GM40; AGTRAP, BRINP2, CDH11, CMTM8, ENSGALG00000001136, FRZB, GADD45A, HUNK, LARP7, LMX1B, MRAS, MSX1, SFRP2, SOX11, SPSB4, Z-FST, ZEB2, ZIC1
# GM42; BMP5, CDH6, CSRNP1, DRAXIN, EN1, ENSGALG00000013505, FOXD3, NKD1, NRP2, OLFM1, PAX7, SNAI2, SOX9, TFAP2B, TMEM132C, TSPAN18, WLS, WNT6, Z-ENC1, ZFHX4, ZNF423
# GM43; COL9A3, ENSGALG00000037717, ENSGALG00000053185, ERMN, ETS1, LMO4, PPP1R1C, RASL11B, RFTN2, SOX10, SOX8, TNC, Z-MEF2C, Z-PLK2
# GM44; CXCR4, ENSGALG00000030512, ENSGALG00000031427, FABP7, FKBP11, GLIPR2, ID1, ID2, MEOX1, MYL4, OLFML3, SOX5, WNT1

NC_genes <- c("AGTRAP", "BRINP2", "CDH11", "CMTM8", "ENSGALG00000001136", "FRZB", "GADD45A", "HUNK", "LARP7", "LMX1B", "MRAS", "MSX1", "SFRP2", "SOX11", "SPSB4", "Z-FST", "ZEB2", "ZIC1",
              "BMP5", "CDH6", "CSRNP1", "DRAXIN", "EN1", "ENSGALG00000013505", "FOXD3", "NKD1", "NRP2", "OLFM1", "PAX7", "SNAI2", "SOX9", "TFAP2B", "TMEM132C", "TSPAN18", "WLS", "WNT6", "Z-ENC1", "ZFHX4", "ZNF423",
              "CDON", "COTL1", "ENSGALG00000048488", "MFAP2", "PRTG", "TUBB3",
              "COL9A3", "ENSGALG00000037717", "ENSGALG00000053185", "ERMN", "ETS1", "LMO4", "PPP1R1C", "RASL11B", "RFTN2", "SOX10", "SOX8", "TNC", "Z-MEF2C", "Z-PLK2",
              "CXCR4", "ENSGALG00000030512", "ENSGALG00000031427", "FABP7", "FKBP11", "GLIPR2", "ID1", "ID2", "MEOX1", "MYL4", "OLFML3", "SOX5", "WNT1")

##############  Read in peaks of interest    ###########################################

## PPR specific PMs: FullData_PM6, FullData_PM7, FullData_PM9
PPR_peaks <- c(
  "chr1-51885355-51885855", "chr1-56743328-56743828", "chr15-2282477-2282977", "chr17-8061757-8062257", "chr2-34364392-34364892", "chr2-72766649-72767149", "chr2-91522834-91523334", "chr25-3194857-3195357", "chr28-2880144-2880644", "chr3-100929473-100929973", "chr3-101800822-101801322", "chr3-18826788-18827288", "chr3-3153527-3154027", "chr3-37532963-37533463", "chr3-40860007-40860507", "chr3-42110813-42111313", "chr3-46021620-46022120", "chr3-9604325-9604825", "chr33-4973016-4973516", "chr4-15228534-15229034", "chr4-3545860-3546360", "chr4-39659304-39659804", "chr4-42955909-42956409", "chr4-61691248-61691748", "chr4-82791681-82792181", "chr5-13025286-13025786", "chr5-14360597-14361097", "chr5-2185340-2185840", "chr5-25039278-25039778", "chr5-26331436-26331936", "chr5-37774715-37775215", "chr6-13570150-13570650", "chr6-22584965-22585465", "chr6-3635106-3635606", "chr7-14213427-14213927", "chr7-26273335-26273835", "chr8-20658815-20659315", "chr8-20919981-20920481", "chr8-8130129-8130629", "chr9-1192259-1192759", "chr9-12061796-12062296",
  "chr1-105919837-105920337", "chr1-113306753-113307253", "chr1-15015808-15016308", "chr1-156920062-156920562", "chr1-157299139-157299639", "chr1-15892762-15893262", "chr1-15919402-15919902", "chr1-168641905-168642405", "chr1-168893830-168894330", "chr1-176740306-176740806", "chr1-194840235-194840735", "chr1-20566555-20567055", "chr1-33571364-33571864", "chr1-60416481-60416981", "chr1-62211336-62211836", "chr1-72723164-72723664", "chr1-8065715-8066215", "chr1-88390949-88391449", "chr1-93486724-93487224", "chr10-14491785-14492285", "chr10-14492580-14493080", "chr10-20340332-20340832", "chr10-20452537-20453037", "chr11-19273365-19273865", "chr11-1987100-1987600", "chr11-2257660-2258160", "chr12-18975182-18975682", "chr12-3599217-3599717", "chr12-6958745-6959245", "chr12-7090763-7091263", "chr12-7158977-7159477", "chr12-7163182-7163682", "chr12-7163824-7164324", "chr12-7292027-7292527", "chr12-8524299-8524799", "chr12-8995625-8996125", "chr13-12548521-12549021", "chr13-16386526-16387026", "chr13-1671556-1672056", "chr13-8986883-8987383", "chr13-9166470-9166970", "chr14-13306078-13306578", "chr14-5259731-5260231", "chr14-5273840-5274340", "chr14-7010284-7010784", "chr15-6045070-6045570", "chr15-767877-768377", "chr17-5823220-5823720", "chr18-3734720-3735220", "chr18-836196-836696", "chr18-9597171-9597671", "chr19-8931349-8931849", "chr2-111543253-111543753", "chr2-113095334-113095834", "chr2-126210787-126211287", "chr2-130112541-130113041", "chr2-147786771-147787271", "chr2-23194456-23194956", "chr2-24093951-24094451", "chr2-37262664-37263164", "chr2-40163142-40163642", "chr2-40666551-40667051", "chr2-40777614-40778114", "chr2-41076879-41077379", "chr2-41096965-41097465", "chr2-46579772-46580272", "chr2-46796604-46797104", "chr2-60768678-60769178", "chr2-90151761-90152261", "chr2-98858224-98858724", "chr20-10297374-10297874", "chr20-3417599-3418099", "chr20-5668390-5668890", "chr20-9882128-9882628", "chr20-9895339-9895839", "chr21-394147-394647", "chr23-2270591-2271091", "chr23-328199-328699", "chr23-4463337-4463837", "chr23-639645-640145", "chr25-2726380-2726880", "chr26-4786832-4787332", "chr26-4822131-4822631", "chr27-4828979-4829479", "chr27-6164738-6165238", "chr27-7146919-7147419", "chr28-2142363-2142863", "chr28-4636074-4636574", "chr3-100893435-100893935", "chr3-102692943-102693443", "chr3-103753008-103753508", "chr3-105402084-105402584", "chr3-108232358-108232858", "chr3-108928367-108928867", "chr3-13375460-13375960", "chr3-1449127-1449627", "chr3-16699856-16700356", "chr3-22710909-22711409", "chr3-32309110-32309610", "chr3-32517550-32518050", "chr3-39074445-39074945", "chr3-39370694-39371194", "chr3-64337111-64337611", "chr3-67485272-67485772", "chr3-71754992-71755492", "chr3-79368221-79368721", "chr3-8650978-8651478", "chr3-87433943-87434443", "chr3-97408796-97409296", "chr33-4665299-4665799", "chr4-1390646-1391146", "chr4-17556200-17556700", "chr4-17640299-17640799", "chr4-240453-240953", "chr4-33759038-33759538", "chr4-34644939-34645439", "chr4-354006-354506", "chr4-43388891-43389391", "chr4-83638368-83638868", "chr4-9476451-9476951", "chr5-1148018-1148518", "chr5-17414692-17415192", "chr5-24251959-24252459", "chr5-3948608-3949108", "chr5-40035538-40036038", "chr7-1009852-1010352", "chr7-33112267-33112767", "chr8-26298399-26298899", "chr9-18127591-18128091",
  "chr1-115742625-115743125", "chr1-118138264-118138764", "chr1-124572961-124573461", "chr1-131433117-131433617", "chr1-140680906-140681406", "chr1-141276478-141276978", "chr1-142414827-142415327", "chr1-1548751-1549251", "chr1-16017632-16018132", "chr1-16510397-16510897", "chr1-166577443-166577943", "chr1-169475933-169476433", "chr1-18484334-18484834", "chr1-186493813-186494313", "chr1-194742462-194742962", "chr1-20918719-20919219", "chr1-24222888-24223388", "chr1-25525043-25525543", "chr1-29787549-29788049", "chr1-29823089-29823589", "chr1-30789661-30790161", "chr1-34361501-34362001", "chr1-34362002-34362502", "chr1-38326892-38327392", "chr1-38366243-38366743", "chr1-39551257-39551757", "chr1-42364379-42364879", "chr1-43485243-43485743", "chr1-46478536-46479036", "chr1-60970364-60970864", "chr1-61696787-61697287", "chr1-85099460-85099960", "chr1-86749742-86750242", "chr1-8841329-8841829", "chr1-96390837-96391337", "chr1-98741966-98742466", "chr10-10272404-10272904", "chr10-12849026-12849526", "chr10-5692798-5693298", "chr10-7462858-7463358", "chr11-15521598-15522098", "chr11-15529701-15530201", "chr11-388210-388710", "chr12-11534333-11534833", "chr12-14776981-14777481", "chr12-17501359-17501859", "chr12-18364771-18365271", "chr12-18943301-18943801", "chr13-12527823-12528323", "chr13-13545910-13546410", "chr13-18191067-18191567", "chr13-18442010-18442510", "chr13-9207099-9207599", "chr14-5275234-5275734", "chr17-5108318-5108818", "chr17-8890725-8891225", "chr18-4077933-4078433", "chr19-9685399-9685899", "chr2-149081179-149081679", "chr2-19630486-19630986", "chr2-39091436-39091936", "chr2-50593634-50594134", "chr2-5316227-5316727", "chr2-63741706-63742206", "chr2-63931517-63932017", "chr20-1015766-1016266", "chr20-11539936-11540436", "chr20-12379867-12380367", "chr20-13165061-13165561", "chr21-1003730-1004230", "chr21-1362953-1363453", "chr21-2402710-2403210", "chr21-2705825-2706325", "chr22-1883318-1883818", "chr23-2249285-2249785", "chr23-4880807-4881307", "chr23-5189129-5189629", "chr26-1704254-1704754", "chr26-618206-618706", "chr28-416980-417480", "chr3-11684665-11685165", "chr3-17132759-17133259", "chr3-353593-354093", "chr3-48440133-48440633", "chr3-65064074-65064574", "chr4-23449835-23450335", "chr4-49589263-49589763", "chr4-50663077-50663577", "chr4-65285432-65285932", "chr4-77824880-77825380", "chr7-8114384-8114884"
)

# Neural specific PMs: Full_Data PMs16, 17, 19, 13, 26, 27, 28, 14, 15, 21, 22
neural_peaks <- c(
  "chr1-106137929-106138429", "chr1-114490138-114490638", "chr1-137546674-137547174", "chr1-142417203-142417703", "chr1-146094796-146095296", "chr1-147839605-147840105", "chr1-16040996-16041496", "chr1-16041588-16042088", "chr1-167012353-167012853", "chr1-167281007-167281507", "chr1-176592352-176592852", "chr1-176717142-176717642", "chr1-182075420-182075920", "chr1-182075980-182076480", "chr1-193639505-193640005", "chr1-193656910-193657410", "chr1-38100825-38101325", "chr1-44448086-44448586", "chr1-61679524-61680024", "chr1-61688631-61689131", "chr1-72190478-72190978", "chr1-82987510-82988010", "chr1-84446364-84446864", "chr10-19858953-19859453", "chr11-10479266-10479766", "chr12-17120301-17120801", "chr15-2267250-2267750", "chr17-10044536-10045036", "chr18-2085934-2086434", "chr2-112876709-112877209", "chr2-122994049-122994549", "chr2-128542303-128542803", "chr2-14051867-14052367", "chr2-22625888-22626388", "chr2-28952935-28953435", "chr2-50690353-50690853", "chr2-58923230-58923730", "chr2-60910724-60911224", "chr2-92516209-92516709", "chr20-314959-315459", "chr20-8604626-8605126", "chr20-9789082-9789582", "chr20-9797880-9798380", "chr21-3905369-3905869", "chr25-3837360-3837860", "chr26-2797110-2797610", "chr3-103859980-103860480", "chr3-106279211-106279711", "chr3-23004031-23004531", "chr3-25954820-25955320", "chr3-43113615-43114115", "chr3-94533991-94534491", "chr33-7469954-7470454", "chr4-31986912-31987412", "chr4-3954823-3955323", "chr4-4788731-4789231", "chr5-23590005-23590505", "chr7-6373594-6374094",
  "chr1-106078312-106078812", "chr1-112594491-112594991", "chr1-113710661-113711161", "chr1-113828534-113829034", "chr1-119272534-119273034", "chr1-141165266-141165766", "chr1-145735373-145735873", "chr1-145749394-145749894", "chr1-145751968-145752468", "chr1-147840212-147840712", "chr1-147841247-147841747", "chr1-15902297-15902797", "chr1-178301745-178302245", "chr1-190296603-190297103", "chr1-194560619-194561119", "chr1-45604616-45605116", "chr1-46338735-46339235", "chr1-49675907-49676407", "chr1-57635404-57635904", "chr1-7709814-7710314", "chr1-93270560-93271060", "chr1-94343375-94343875", "chr13-11401600-11402100", "chr15-6204062-6204562", "chr15-7634258-7634758", "chr17-10163514-10164014", "chr17-3628784-3629284", "chr17-4595340-4595840", "chr17-9580542-9581042", "chr18-10708567-10709067", "chr18-4018799-4019299", "chr18-5405570-5406070", "chr18-9155220-9155720", "chr19-4773088-4773588", "chr19-5512393-5512893", "chr19-5625858-5626358", "chr19-7949216-7949716", "chr2-14475402-14475902", "chr2-17241389-17241889", "chr2-17355198-17355698", "chr2-23488883-23489383", "chr2-64762558-64763058", "chr20-11025130-11025630", "chr20-3795814-3796314", "chr20-8142146-8142646", "chr20-8606486-8606986", "chr22-1117247-1117747", "chr23-433251-433751", "chr24-576054-576554", "chr25-3369637-3370137", "chr27-5813457-5813957", "chr28-3551922-3552422", "chr3-18188117-18188617", "chr3-30353331-30353831", "chr3-52863504-52864004", "chr3-86481773-86482273", "chr4-20864786-20865286", "chr4-31499500-31500000", "chr4-582733-583233", "chr5-36884277-36884777", "chr5-8473833-8474333",
  "chr1-104105509-104106009", "chr1-105869073-105869573", "chr1-121390111-121390611", "chr1-125902526-125903026", "chr1-134959165-134959665", "chr1-139692500-139693000", "chr1-14400380-14400880", "chr1-146605876-146606376", "chr1-175178638-175179138", "chr1-176596186-176596686", "chr1-178764388-178764888", "chr1-186266458-186266958", "chr1-187677804-187678304", "chr1-194564410-194564910", "chr1-34154397-34154897", "chr1-67008470-67008970", "chr1-68751180-68751680", "chr10-12281366-12281866", "chr10-2830982-2831482", "chr11-11274849-11275349", "chr11-18655061-18655561", "chr11-19351127-19351627", "chr11-8613190-8613690", "chr12-13914734-13915234", "chr12-1790389-1790889", "chr12-19386518-19387018", "chr12-20036679-20037179", "chr12-2977540-2978040", "chr12-3299978-3300478", "chr13-11403340-11403840", "chr14-9323907-9324407", "chr15-4254305-4254805", "chr15-5401335-5401835", "chr15-8926325-8926825", "chr17-4267538-4268038", "chr17-9600748-9601248", "chr18-2362318-2362818", "chr18-4888689-4889189", "chr18-7042454-7042954", "chr19-1580191-1580691", "chr19-4373319-4373819", "chr2-110546586-110547086", "chr2-116482267-116482767", "chr2-117559250-117559750", "chr2-11828007-11828507", "chr2-127708825-127709325", "chr2-128567844-128568344", "chr2-129097316-129097816", "chr2-142016911-142017411", "chr2-14570544-14571044", "chr2-146267996-146268496", "chr2-17404120-17404620", "chr2-24450002-24450502", "chr2-68523385-68523885", "chr2-80845764-80846264", "chr2-97101828-97102328", "chr20-13679070-13679570", "chr20-5119410-5119910", "chr20-8143191-8143691", "chr20-961903-962403", "chr21-1734148-1734648", "chr21-3180813-3181313", "chr21-499814-500314", "chr21-586259-586759", "chr23-2308823-2309323", "chr23-5503140-5503640", "chr23-615551-616051", "chr24-3548265-3548765", "chr24-4965487-4965987", "chr25-2554756-2555256", "chr27-4137401-4137901", "chr3-105686641-105687141", "chr3-77319621-77320121", "chr4-3617735-3618235", "chr4-43213175-43213675", "chr4-87879949-87880449", "chr5-54425373-54425873", "chr5-58514315-58514815", "chr6-24014438-24014938", "chr7-7249739-7250239", "chr9-15920368-15920868",
  "chr1-195854609-195855109", "chr1-437304-437804", "chr11-16396018-16396518", "chr15-10869732-10870232", "chr15-2979865-2980365", "chr15-640016-640516", "chr18-2360860-2361360", "chr2-24110502-24111002", "chr2-30172317-30172817", "chr2-30770975-30771475", "chr2-31222184-31222684", "chr2-33953299-33953799", "chr2-48408615-48409115", "chr20-10555339-10555839", "chr20-4949645-4950145", "chr21-3822831-3823331", "chr23-3659418-3659918", "chr23-5072895-5073395", "chr23-5107466-5107966", "chr23-5606308-5606808", "chr26-4847291-4847791", "chr28-3138091-3138591", "chr3-27103783-27104283", "chr3-36774909-36775409", "chr3-57056055-57056555", "chr3-78190622-78191122", "chr3-79546632-79547132", "chr3-82655389-82655889", "chr3-83535943-83536443", "chr3-9552521-9553021", "chr3-96223021-96223521", "chr3-97422967-97423467", "chr3-98103065-98103565", "chr30-11290-11790", "chr33-7507205-7507705", "chr4-10383274-10383774", "chr4-10661555-10662055", "chr4-15669597-15670097", "chr4-17184069-17184569", "chr4-18644781-18645281", "chr4-39170183-39170683", "chr4-39688374-39688874", "chr4-40451153-40451653", "chr4-44577557-44578057", "chr4-46740151-46740651", "chr4-53123977-53124477", "chr4-54159409-54159909", "chr4-69966399-69966899", "chr4-74522333-74522833", "chr4-89894073-89894573", "chr4-9317467-9317967", "chr5-13332551-13333051", "chr5-14649722-14650222", "chr5-24586878-24587378", "chr5-27825787-27826287", "chr5-28251517-28252017", "chr5-3636251-3636751", "chr5-3702738-3703238", "chr5-43585154-43585654", "chr5-43589273-43589773", "chr5-4832033-4832533", "chr5-52765303-52765803", "chr5-54399939-54400439", "chr5-56209103-56209603", "chr5-56230653-56231153", "chr5-56243559-56244059", "chr5-58793925-58794425", "chr5-59465059-59465559", "chr6-12671953-12672453", "chr6-12676908-12677408", "chr6-18140596-18141096", "chr6-28183645-28184145", "chr6-28845291-28845791", "chr6-4041964-4042464", "chr6-6837521-6838021", "chr6-8627902-8628402", "chr6-9942185-9942685", "chr7-11774670-11775170", "chr7-13095261-13095761", "chr7-22405342-22405842", "chr7-23236796-23237296", "chr7-23906742-23907242", "chr7-26211823-26212323", "chr7-27395063-27395563", "chr7-27764823-27765323", "chr7-32340504-32341004", "chr7-32377264-32377764", "chr7-35367195-35367695", "chr8-14776909-14777409", "chr8-16015389-16015889", "chr8-17917230-17917730", "chr8-22259912-22260412", "chr8-26324710-26325210", "chr8-26942711-26943211", "chr8-27055168-27055668", "chr8-29567839-29568339", "chr9-14828282-14828782", "chr9-15572398-15572898", "chr9-2678927-2679427", "chr9-6774452-6774952", "chr9-9382177-9382677",
  "chr1-100256686-100257186", "chr1-116717709-116718209", "chr1-125639291-125639791", "chr1-132728509-132729009", "chr1-133746833-133747333", "chr1-137091574-137092074", "chr1-144786422-144786922", "chr1-146480625-146481125", "chr1-154047021-154047521", "chr1-155323057-155323557", "chr1-156588427-156588927", "chr1-161584951-161585451", "chr1-164455937-164456437", "chr1-169647146-169647646", "chr1-180122422-180122922", "chr1-190186765-190187265", "chr1-194277569-194278069", "chr1-25500877-25501377", "chr1-35760261-35760761", "chr1-38494594-38495094", "chr1-45591945-45592445", "chr1-48317495-48317995", "chr1-55047675-55048175", "chr1-62025911-62026411", "chr1-81068901-81069401", "chr1-88366175-88366675", "chr1-9351165-9351665", "chr1-9504434-9504934", "chr10-11137090-11137590", "chr10-18333438-18333938", "chr11-10527202-10527702", "chr11-1224313-1224813", "chr11-14096883-14097383", "chr11-2273384-2273884", "chr11-6869715-6870215", "chr11-9307918-9308418", "chr12-3636024-3636524", "chr12-5379473-5379973", "chr13-13319874-13320374", "chr13-14602419-14602919", "chr13-2772793-2773293", "chr13-9813918-9814418", "chr14-10548959-10549459", "chr14-13453167-13453667", "chr14-14951186-14951686", "chr14-5567618-5568118", "chr14-7432390-7432890", "chr14-8440448-8440948", "chr15-5819389-5819889", "chr17-3522245-3522745", "chr17-4278102-4278602", "chr17-8544632-8545132", "chr18-2417132-2417632", "chr18-3974002-3974502", "chr18-9217798-9218298", "chr18-9500788-9501288", "chr18-9635587-9636087", "chr19-3410659-3411159", "chr19-9546854-9547354", "chr19-9864558-9865058", "chr2-102409084-102409584", "chr2-103177320-103177820", "chr2-104115015-104115515", "chr2-105730649-105731149", "chr2-107417751-107418251", "chr2-112543002-112543502", "chr2-114807899-114808399", "chr2-126309163-126309663", "chr2-128491470-128491970", "chr2-128514400-128514900", "chr2-128520410-128520910", "chr2-131301746-131302246", "chr2-131521254-131521754", "chr2-131986026-131986526", "chr2-146317757-146318257", "chr2-16418565-16419065", "chr2-17348515-17349015", "chr2-18398171-18398671", "chr2-24646983-24647483", "chr2-52931398-52931898", "chr2-5757473-5757973", "chr2-60260978-60261478", "chr2-60826084-60826584", "chr2-61073871-61074371", "chr2-67931588-67932088", "chr2-68500628-68501128", "chr2-76660469-76660969", "chr2-81852400-81852900", "chr2-89417428-89417928", "chr2-90053501-90054001", "chr2-90154735-90155235", "chr2-98511194-98511694", "chr20-3434671-3435171", "chr20-5118683-5119183", "chr21-504539-505039", "chr24-4326078-4326578", "chr26-3360986-3361486", "chr3-38162601-38163101", "chr3-64800955-64801455", "chr3-93907144-93907644", "chr4-20053949-20054449", "chr4-49545592-49546092", "chr4-56547912-56548412",
  "chr1-31897857-31898357", "chr10-18450894-18451394", "chr11-11636484-11636984", "chr11-12122002-12122502", "chr11-17881110-17881610", "chr11-8439497-8439997", "chr12-17027688-17028188", "chr12-8849446-8849946", "chr12-9076346-9076846", "chr13-13146241-13146741", "chr13-16941293-16941793", "chr13-16946839-16947339", "chr13-1807903-1808403", "chr14-9231421-9231921", "chr14-9649903-9650403", "chr17-5186326-5186826", "chr17-6290577-6291077", "chr19-585888-586388", "chr2-102409609-102410109", "chr2-127499763-127500263", "chr2-128290869-128291369", "chr2-144151666-144152166", "chr2-64724389-64724889", "chr2-98449420-98449920", "chr3-31894203-31894703", "chr3-44227478-44227978", "chr3-47478850-47479350", "chr3-54393384-54393884", "chr3-67859141-67859641", "chr4-32142294-32142794", "chr4-5295545-5296045", "chr4-5319180-5319680", "chr4-82075520-82076020", "chr5-380020-380520", "chr5-44850349-44850849",
  "chr1-113734319-113734819", "chr1-139468955-139469455", "chr1-155149185-155149685", "chr1-166957157-166957657", "chr1-176652448-176652948", "chr1-186263202-186263702", "chr1-193246244-193246744", "chr1-24619813-24620313", "chr1-35500578-35501078", "chr1-45612573-45613073", "chr1-48558236-48558736", "chr1-50576984-50577484", "chr10-6613240-6613740", "chr10-8243546-8244046", "chr13-10757672-10758172", "chr13-9656040-9656540", "chr14-15051284-15051784", "chr14-2938297-2938797", "chr15-7394568-7395068", "chr17-2787493-2787993", "chr17-3255673-3256173", "chr17-4540664-4541164", "chr17-6890686-6891186", "chr17-7583467-7583967", "chr18-3946716-3947216", "chr19-359660-360160", "chr19-9113103-9113603", "chr2-127294606-127295106", "chr2-130283571-130284071", "chr2-132626292-132626792", "chr2-22099742-22100242", "chr2-22747752-22748252", "chr2-25562172-25562672", "chr2-40078646-40079146", "chr2-41432147-41432647", "chr2-45362270-45362770", "chr2-60965537-60966037", "chr2-74678147-74678647", "chr2-7777356-7777856", "chr2-92020457-92020957", "chr20-10764886-10765386", "chr20-653122-653622", "chr21-1971695-1972195", "chr21-2251894-2252394", "chr21-4039469-4039969", "chr21-4750491-4750991", "chr21-5787445-5787945", "chr22-2484756-2485256", "chr23-1036392-1036892", "chr26-4796303-4796803", "chr27-2780042-2780542", "chr28-4242432-4242932", "chr3-105488111-105488611", "chr3-2684843-2685343", "chr3-31874282-31874782", "chr3-33328841-33329341", "chr3-38402855-38403355", "chr3-60140116-60140616", "chr3-60210469-60210969", "chr3-87719428-87719928", "chr3-87733453-87733953", "chr3-88399347-88399847", "chr3-9096314-9096814", "chr3-95165591-95166091", "chr4-10141669-10142169", "chr4-22534112-22534612", "chr4-51494974-51495474", "chr5-10610208-10610708", "chr5-10977739-10978239", "chr5-13257396-13257896", "chr5-14357450-14357950", "chr5-16549438-16549938", "chr5-22419185-22419685", "chr5-34423373-34423873", "chr5-35629540-35630040", "chr6-12261549-12262049", "chr6-14163485-14163985", "chr7-18755118-18755618",
  "chr1-124277125-124277625", "chr1-145879399-145879899", "chr1-146481262-146481762", "chr1-147742303-147742803", "chr1-154320116-154320616", "chr1-166988277-166988777", "chr1-177913122-177913622", "chr1-186438192-186438692", "chr1-19933989-19934489", "chr1-54033672-54034172", "chr1-56660357-56660857", "chr1-56667243-56667743", "chr1-59415654-59416154", "chr1-6425003-6425503", "chr1-77220638-77221138", "chr1-82718315-82718815", "chr1-84422819-84423319", "chr1-93745097-93745597", "chr10-6584568-6585068", "chr12-19034683-19035183", "chr12-6097894-6098394", "chr12-6421552-6422052", "chr12-9251683-9252183", "chr14-9193837-9194337", "chr15-4589617-4590117", "chr15-7385842-7386342", "chr17-10083032-10083532", "chr18-5969536-5970036", "chr2-114261826-114262326", "chr2-12576862-12577362", "chr2-131813649-131814149", "chr2-142673798-142674298", "chr2-16305749-16306249", "chr2-77651262-77651762", "chr2-88781629-88782129", "chr2-98745135-98745635", "chr21-6315097-6315597", "chr22-2002127-2002627", "chr23-1241163-1241663", "chr23-3662649-3663149", "chr26-2249143-2249643", "chr3-24867448-24867948", "chr3-59657330-59657830", "chr4-19811034-19811534", "chr4-54219883-54220383", "chr5-59642723-59643223", "chr6-17360644-17361144",
  "chr1-179889954-179890454", "chr1-182092325-182092825", "chr1-186339282-186339782", "chr1-29544379-29544879", "chr1-29849959-29850459", "chr1-34010511-34011011", "chr1-3858483-3858983", "chr1-65910069-65910569", "chr1-69990169-69990669", "chr10-13554754-13555254", "chr10-6550197-6550697", "chr10-8184958-8185458", "chr11-15423997-15424497", "chr11-6917350-6917850", "chr13-12489033-12489533", "chr14-1981397-1981897", "chr17-5386529-5387029", "chr2-149508318-149508818", "chr2-37179732-37180232", "chr2-54941239-54941739", "chr2-67048366-67048866", "chr20-11519134-11519634", "chr20-4535098-4535598", "chr21-6045272-6045772", "chr24-4643636-4644136", "chr26-554564-555064", "chr27-4159958-4160458", "chr27-6767843-6768343", "chr28-1804307-1804807", "chr28-2404605-2405105", "chr3-2514837-2515337", "chr3-56167473-56167973", "chr3-62752110-62752610", "chr3-64643199-64643699", "chr3-70965987-70966487", "chr3-87462753-87463253", "chr4-579496-579996", "chr4-63719457-63719957", "chr4-84146336-84146836", "chr5-12594349-12594849", "chr5-15651436-15651936", "chr8-4244120-4244620",
  "chr1-7437076-7437576", "chr10-11587997-11588497", "chr10-3279675-3280175", "chr10-4132717-4133217", "chr10-6860577-6861077", "chr11-1108100-1108600", "chr11-12809191-12809691", "chr11-4255498-4255998", "chr11-8587183-8587683", "chr12-2970476-2970976", "chr13-16294316-16294816", "chr13-6376462-6376962", "chr14-11081634-11082134", "chr14-7150109-7150609", "chr19-3846849-3847349", "chr2-21927124-21927624", "chr2-26031395-26031895", "chr2-50806440-50806940", "chr3-107956928-107957428", "chr4-52022428-52022928",
  "chr1-181028374-181028874", "chr1-182195411-182195911", "chr1-189094912-189095412", "chr1-38094574-38095074", "chr1-4030993-4031493", "chr1-65800555-65801055", "chr1-65823168-65823668", "chr1-65849041-65849541", "chr1-7556848-7557348", "chr1-84462615-84463115", "chr1-88810768-88811268", "chr10-5228935-5229435", "chr10-6862897-6863397", "chr18-3979856-3980356", "chr19-5216693-5217193", "chr2-106497272-106497772", "chr2-16742368-16742868", "chr2-40625009-40625509", "chr2-65411594-65412094", "chr2-85768502-85769002", "chr2-88181852-88182352", "chr20-7533237-7533737", "chr20-9647147-9647647", "chr3-64431963-64432463", "chr3-88544425-88544925", "chr33-7095286-7095786", "chr4-21371467-21371967", "chr4-80375145-80375645", "chr5-55744743-55745243", "chr8-12381541-12382041"
)

# Epiblast PMs: Full_Data PMs2, 3, 8
epiblast_peaks <- c(
  "chr10-16877201-16877701", "chr10-20305630-20306130", "chr10-7846100-7846600", "chr11-1262663-1263163", "chr11-6689169-6689669", "chr12-17709894-17710394", "chr12-17984023-17984523", "chr12-18880567-18881067", "chr12-19263478-19263978", "chr12-2011549-2012049", "chr12-2893244-2893744", "chr12-3588217-3588717", "chr13-11096560-11097060", "chr13-11200008-11200508", "chr13-12302020-12302520", "chr13-16261807-16262307", "chr13-16317445-16317945", "chr13-16331559-16332059", "chr13-18245308-18245808", "chr13-18616212-18616712", "chr13-18623581-18624081", "chr13-2059792-2060292", "chr13-9664712-9665212", "chr13-9668244-9668744", "chr14-12910705-12911205", "chr14-1568588-1569088", "chr14-3268190-3268690", "chr14-5769533-5770033", "chr14-6105301-6105801", "chr14-7459631-7460131", "chr14-8807413-8807913",
  "chr1-109784285-109784785", "chr1-117188382-117188882", "chr1-119408593-119409093", "chr1-121259653-121260153", "chr1-126605738-126606238", "chr1-127044148-127044648", "chr1-154056799-154057299", "chr1-157424727-157425227", "chr1-157425253-157425753", "chr1-159079303-159079803", "chr1-16065388-16065888", "chr1-173971901-173972401", "chr1-177125052-177125552", "chr1-177397329-177397829", "chr1-25342104-25342604", "chr1-27043376-27043876", "chr1-39810582-39811082", "chr1-84352601-84353101", "chr1-86784143-86784643", "chr1-92139743-92140243", "chr1-92667713-92668213", "chr10-18004931-18005431", "chr10-9146604-9147104", "chr10-9630113-9630613", "chr13-13513141-13513641", "chr14-5173490-5173990", "chr15-2243337-2243837", "chr17-6750892-6751392", "chr17-8979427-8979927", "chr17-987138-987638", "chr18-4208895-4209395", "chr18-7795710-7796210", "chr19-2006748-2007248", "chr19-7913523-7914023", "chr19-8116052-8116552", "chr19-9283211-9283711", "chr2-101226221-101226721", "chr2-107644909-107645409", "chr2-107646386-107646886", "chr2-107721543-107722043", "chr2-111084665-111085165", "chr2-114807199-114807699", "chr2-115047514-115048014", "chr2-115084873-115085373", "chr2-115095547-115096047", "chr2-115132641-115133141", "chr2-115907429-115907929", "chr2-117719644-117720144", "chr2-117769270-117769770", "chr2-118690943-118691443", "chr2-125108629-125109129", "chr2-126371612-126372112", "chr2-135035551-135036051", "chr2-17094137-17094637", "chr2-20591570-20592070", "chr2-21520558-21521058", "chr2-2385823-2386323", "chr2-28625972-28626472", "chr2-41399688-41400188", "chr2-4147053-4147553", "chr2-44093548-44094048", "chr2-4456308-4456808", "chr2-44731879-44732379", "chr2-61296837-61297337", "chr2-6477136-6477636", "chr2-64970720-64971220", "chr2-75478801-75479301", "chr2-8168334-8168834", "chr2-85019745-85020245", "chr2-85390201-85390701", "chr2-8834638-8835138", "chr2-89602516-89603016", "chr2-96625859-96626359", "chr20-4343423-4343923", "chr21-2731232-2731732", "chr21-3042775-3043275", "chr21-3140293-3140793", "chr21-4379846-4380346", "chr21-866341-866841", "chr22-1435313-1435813", "chr23-2493700-2494200", "chr23-3786433-3786933", "chr24-2318649-2319149", "chr26-4541909-4542409", "chr26-4542507-4543007", "chr26-4545177-4545677", "chr27-7030870-7031370", "chr3-52039292-52039792", "chr3-56200142-56200642", "chr3-65070508-65071008", "chr3-87615531-87616031", "chr3-90980485-90980985", "chr4-12226935-12227435", "chr4-19731560-19732060", "chr4-48152913-48153413", "chr4-5370874-5371374", "chr4-59717236-59717736",
  "chr1-110281882-110282382", "chr11-8925624-8926124", "chr14-3706879-3707379", "chr18-1600943-1601443", "chr18-5994505-5995005", "chr18-9739470-9739970", "chr19-8482522-8483022", "chr2-106895115-106895615", "chr2-113123525-113124025", "chr2-114269644-114270144", "chr2-126301655-126302155", "chr2-142451443-142451943", "chr2-38837219-38837719", "chr2-45419343-45419843", "chr2-68174576-68175076", "chr2-73341546-73342046", "chr2-74258115-74258615", "chr2-78063433-78063933", "chr2-78256853-78257353", "chr2-78556854-78557354", "chr2-79869348-79869848", "chr2-80039828-80040328", "chr2-81301583-81302083", "chr2-85765249-85765749", "chr2-87799837-87800337", "chr2-89608377-89608877", "chr20-1215812-1216312", "chr20-2435081-2435581", "chr21-1425482-1425982", "chr22-1221930-1222430", "chr22-1223279-1223779", "chr22-2694735-2695235", "chr24-2230342-2230842", "chr26-4429645-4430145", "chr28-3743828-3744328", "chr3-103855082-103855582", "chr3-107043466-107043966", "chr3-109761452-109761952", "chr3-27313048-27313548", "chr3-38464922-38465422", "chr3-51455828-51456328", "chr3-65736216-65736716", "chr3-7676617-7677117", "chr3-91321007-91321507", "chr3-91418181-91418681", "chr4-1092590-1093090", "chr4-1166664-1167164", "chr4-32359369-32359869", "chr4-44150493-44150993", "chr4-49797480-49797980", "chr4-553198-553698", "chr4-82536788-82537288", "chr5-22208723-22209223", "chr7-9765533-9766033"
)


##############  Look at which interactions include PPR genes    ###########################################


## should write a function which:
# inputs 1 or more features (genes or peaks)
# extracts the bins which overlaps with the input feature(s)
# extracts the interactions which overlap with the bins
# extracts the OTHER bins which these interactions connect with (ie not input bin)
# returns a table like this:
# input feature | input bin | interaction | other bin | other bin feature | other bin feature type


########################
## function to extract bins from features (gene_names, gene_IDs or peak_IDs)
extract_bins_from_features <- function(features, feature_type, bin_dictionary){
  if (!feature_type %in% c("gene_name", "gene_ID", "peak_ID")) {
    stop("feature_type must be 'gene_name', 'gene_ID' or 'peak_ID'!")
  } else {
    bins <- bin_dictionary %>% 
      dplyr::filter(get(feature_type) %in% features)
  }
  if (feature_type == "peak_ID"){
    bins <- bins %>% dplyr::select(peak_ID, bin_ID)
  } else {
    bins <- bins %>% dplyr::select(gene_ID, gene_name, bin_ID)
  }
  return(bins)
}
#############

##############
##### reads in input bins and finds interactions
## function to extract all bins that interact with input bins
# removes bins which are not interacting significantly
# annotated anchors of interactions with the information from the input bins
## input_bins is bins to search for, input_feature_type is the type of annotation of input bins
## interactions is the dictionary of interactions to search
extract_interacting_bins <- function(input_bins, input_feature_type, interactions){
  
  if (!input_feature_type %in% c("gene", "peak")) {
    stop("input_feature_type must be 'gene', 'peak'!")
  }
  
  # extract interactions with the input bins
  extracted_interactions <- interactions %>% 
    filter(anchor_I %in% input_bins$bin_ID | anchor_J %in% input_bins$bin_ID) %>%
    dplyr::select(anchor_I, anchor_J)
  
  # merge the input bins annotations with the anchors
  if(input_feature_type == "gene"){
    colnames(input_bins) <- c("anchor_I_gene_ID", "anchor_I_gene_name", "anchor_I")
    output_interactions <- merge(extracted_interactions, input_bins, by = "anchor_I", all.x = TRUE)
    colnames(input_bins) <- c("anchor_J_gene_ID", "anchor_J_gene_name", "anchor_J")
    output_interactions <- merge(output_interactions, input_bins, by = "anchor_J", all.x = TRUE)
  }
  if(input_feature_type == "peak"){
    colnames(input_bins) <- c("anchor_I_peak_ID", "anchor_I")
    output_interactions <- merge(extracted_interactions, input_bins, by = "anchor_I", all.x = TRUE)
    colnames(input_bins) <- c("anchor_J_peak_ID", "anchor_J")
    output_interactions <- merge(output_interactions, input_bins, by = "anchor_J", all.x = TRUE)
  }
  
  # identify whether the input anchors are in anchor I, anchor J or both
  output_interactions <- output_interactions %>%
    mutate(input_anchor = case_when(
      !is.na(anchor_I_peak_ID) & !is.na(anchor_J_peak_ID) ~ "both",
      !is.na(anchor_I_peak_ID) ~ "anchor_I",
      !is.na(anchor_J_peak_ID) ~ "anchor_J",
      TRUE ~ NA_character_
    ))
  # then swap around the anchors (they dont mean anything anyway) so all input anchors are anchor I
  output_interactions[output_interactions$input_anchor == "anchor_J", c("anchor_J", "anchor_I")] <- output_interactions[output_interactions$input_anchor == "anchor_J", c("anchor_I", "anchor_J")]
  output_interactions[output_interactions$input_anchor == "anchor_J", c("anchor_I_peak_ID", "anchor_J_peak_ID")] <- output_interactions[output_interactions$input_anchor == "anchor_J", c("anchor_J_peak_ID", "anchor_I_peak_ID")] 
  output_interactions <- output_interactions[,-5]
  
  return(output_interactions)
}
##############

########################
## function to extract features from binIDs
# output is collapsed so each row = binID, annotated with all interacting peaks and promoters
extract_features_from_bins <- function(output_bins, peak_bin_dictionary, promoter_bin_dictionary){
  # extract all peaks in these bins and collapse
  peaks <- peak_bin_dictionary %>% 
    dplyr::filter(peak_bin_dictionary$bin_ID %in% output_bins$anchor_I | peak_bin_dictionary$bin_ID %in% output_bins$anchor_J) %>%
    dplyr::select(peak_ID, bin_ID) %>%
    group_by(bin_ID) %>%
    summarise(across(everything(), ~paste0(unique(.), collapse = ",")))
  # extract all gene promoters in these bins and collapse
  promoters <- promoter_bin_dictionary %>%
    dplyr::filter(promoter_bin_dictionary$bin_ID %in% output_bins$anchor_I | promoter_bin_dictionary$bin_ID %in% output_bins$anchor_J) %>%
    dplyr::select(gene_ID, gene_name, bin_ID) %>%
    group_by(bin_ID) %>%
    summarise(across(everything(), ~paste0(unique(.), collapse = ",")))
  # add peak annotations to the anchors in the interactions
  colnames(peaks) <- c("anchor_I", "anchor_I_peak_ID")
  output_interactions <- merge(output_bins, peaks, by = "anchor_I", all.x = TRUE)
  colnames(peaks) <- c("anchor_J", "anchor_J_peak_ID")
  output_interactions <- merge(output_interactions, peaks, by = "anchor_J", all.x = TRUE)
  # add gene annotations to the anchors in the interactions
  colnames(promoters) <- c("anchor_I", "anchor_I_gene_ID", "anchor_I_gene_name")
  output_interactions <- merge(output_interactions, promoters, by = "anchor_I", all.x = TRUE)
  colnames(promoters) <- c("anchor_J", "anchor_J_gene_ID", "anchor_J_gene_name")
  output_interactions <- merge(output_interactions, promoters, by = "anchor_J", all.x = TRUE)
  return(output_interactions)
}

####### Running through processes from peaks to genes:
## for PPR PMs 6, 7, 9 (should connect with GM23 on full data)

# 1) Extract input_bins from PPR peaks
input_bins <- extract_bins_from_features(features=PPR_peaks, feature_type="peak_ID", bin_dictionary=peaks_bins)

# 2) See which output_bins are interacting with these input_bins
output_bins <- extract_interacting_bins(input_bins = input_bins, input_feature_type = "peak", interactions = filtered_interactions)

# 3) See which features are in these output_bins
output_features <- extract_features_from_bins(output_bins, peak_bin_dictionary = peaks_bins, promoter_bin_dictionary = promoters_bins)

# 4) Investigate overlap of interacting genes with gene modules
interacting_genes <- unique(na.omit(output_features$anchor_J_gene_name))
length(interacting_genes)

GM23 <- c("ASS1", "BMP4", "BMP6", "CLDN3", "CSRP2", "DLX5", "DLX6", "EMILIN2", "ENSGALG00000001885", "ENSGALG00000023936", "ENSGALG00000040010", "ENSGALG00000042443", "ENSGALG00000050334", "ENSGALG00000052786", "EYA2", "FABP3", "FAM184B", "FAM89A", "FN1", "GATA2", "GATA3", "HAS2", "KCNAB1", "KRT18", "KRT19", "KRT7", "LAMB1", "NEDD9", "NET1", "PITX1", "SIX1", "SPON1", "TFAP2A", "TUBAL3", "UNC5B", "Z-HAPLN1")
sum(interacting_genes %in% GM23)
interacting_genes[interacting_genes %in% GM23]

sum(interacting_genes %in% PPR_genes)
interacting_genes[interacting_genes %in% PPR_genes]

write.csv(output_features, file = paste0(rds_path, "interactions_from_FullDataPM6_7_9.csv"))

####### Running through processes from peaks to genes:
## for neural PMs (should connect with GM9 on full data)

# 1) Extract input_bins from PPR peaks
input_bins <- extract_bins_from_features(features=neural_peaks, feature_type="peak_ID", bin_dictionary=peaks_bins)

# 2) See which output_bins are interacting with these input_bins
output_bins <- extract_interacting_bins(input_bins = input_bins, input_feature_type = "peak", interactions = filtered_interactions)

# 3) See which features are in these output_bins
output_features <- extract_features_from_bins(output_bins, peak_bin_dictionary = peaks_bins, promoter_bin_dictionary = promoters_bins)

# 4) Investigate overlap of interacting genes with gene modules
interacting_genes <- unique(na.omit(output_features$anchor_J_gene_name))
length(interacting_genes)

GM9 <- c("CDH2", "CNTNAP1", "ENSGALG00000054487", "FEZF2", "LMO1", "LRRN1", "MSN", "MYLK", "NUAK1", "SOX21", "TOX", "Z-GRP", "Z-RAX")
sum(interacting_genes %in% GM9)
interacting_genes[interacting_genes %in% GM9]

write.csv(output_features, file = paste0(rds_path, "interactions_from_FullDataPM16_17_19_13_26_27_28_14_15_21_22.csv"))

####### Running through processes from peaks to genes:
## for epiblast PMs (should connect with GM21 and GM24 on full data)

# 1) Extract input_bins from PPR peaks
input_bins <- extract_bins_from_features(features=epiblast_peaks, feature_type="peak_ID", bin_dictionary=peaks_bins)

# 2) See which output_bins are interacting with these input_bins
output_bins <- extract_interacting_bins(input_bins = input_bins, input_feature_type = "peak", interactions = filtered_interactions)

# 3) See which features are in these output_bins
output_features <- extract_features_from_bins(output_bins, peak_bin_dictionary = peaks_bins, promoter_bin_dictionary = promoters_bins)

# 4) Investigate overlap of interacting genes with gene modules
interacting_genes <- unique(na.omit(output_features$anchor_J_gene_name))
length(interacting_genes)

GM21 <- c("ACSL1", "ACTN1", "AKR1D1", "ATP1A1", "ATP1B1", "B3GNT7", "BAMBI", "CCDC25", "CDK6", "CITED4", "CTSB", "CXCL12", "DAG1", "ENSGALG00000002988", "ENSGALG00000004814", "ENSGALG00000005572", "ENSGALG00000008518", "ENSGALG00000016570", "ENSGALG00000027805", "ENSGALG00000051984", "FIBIN", "HOMER2", "ID3", "IRF6", "KLF3", "LY6E", "MAP7", "METRNL", "MYL3", "MYL9", "NDRG1", "NT5DC2", "PDZK1IP1", "RGN", "S100A11", "SMIM4", "TFAP2C", "TTC39A")
GM24 <- c("ADD3", "AIG1", "AMY1A", "AP1S3", "ARFGAP1", "ARL8BL", "ATP2B1", "CACNG3", "CD24", "CLDN1", "ENSGALG00000026754", "ENSGALG00000040886", "ENSGALG00000043421", "EPCAM", "G3BP1", "HSPA9", "IVNS1ABP", "MPZL3", "PDGFA", "SALL4", "SLC25A4", "SUCLG1", "TSPAN13", "Z-CTSV", "Z-PLPP1")
sum(interacting_genes %in% c(GM21, GM24))
sum(c("EXOSC8", "ALG5", "TMEM173", "UBE2D2", "TMEM130", "TRRAP", "DLEC1", "RF00026", "PANK4", "MTOR", "UBIAD1", "STAR", "LSM1", "BAG4", "NSD3", "LETM2", "OARD1", "NYFA", "MCFD2", "TTC7A") %in% c(GM21, GM24))

write.csv(output_features, file = paste0(rds_path, "interactions_from_FullDataPM2_3_8.csv"))

#####################################################################################
#####################################################################################

# ########################
# ##### reads in interactions from gene bins and annotates to promoters
# ## function to extract features from binIDs
# # output is collapsed so each row = binID, annotated with all interacting peaks and promoters
# extract_peaks_from_interactions <- function(output_bins, peak_bin_dictionary){
#   # extract all peaks in these bins and collapse
#   peaks <- peak_bin_dictionary %>% 
#     dplyr::filter(peak_bin_dictionary$bin_ID %in% output_bins$anchor_I | peak_bin_dictionary$bin_ID %in% output_bins$anchor_J) %>%
#     dplyr::select(peak_ID, bin_ID) %>%
#     group_by(bin_ID) %>%
#     summarise(across(everything(), ~paste0(unique(.), collapse = ",")))
#   # add peak annotations to the anchors in the interactions
#   colnames(peaks) <- c("anchor_I", "anchor_I_peak_ID")
#   output_interactions <- merge(output_bins, peaks, by = "anchor_I", all.x = TRUE)
#   colnames(peaks) <- c("anchor_J", "anchor_J_peak_ID")
#   output_interactions <- merge(output_interactions, peaks, by = "anchor_J", all.x = TRUE)
#   return(output_interactions)
# }





# ####### Running through processes from genes to peaks:
# 
# # 1) Extract input_bins from PPR genes
# input_bins <- extract_bins_from_features(features=PPR_genes, feature_type="gene_name", bin_dictionary=promoters_bins)
# 
# # 2) See which output_bins are interacting with these input_bins
# output_bins <- extract_interacting_bins(input_bins = input_bins, interactions = filtered_interactions)
# 
# # 3) See which features are in these output_bins
# output_features <- extract_peaks_from_interactions(output_bins, peaks_bins)
# 
# # 4) See which of the two anchors includes the input genes
# output_features <- output_features %>% mutate(input_gene_id = coalesce(anchor_I_gene_ID, anchor_J_gene_ID)) %>%
#   mutate(input_gene_name = coalesce(anchor_I_gene_name, anchor_J_gene_name)) %>% 
#   select(anchor_I, anchor_J, input_gene_id, input_gene_name, anchor_I_peak_ID, anchor_J_peak_ID)
# 
# # Extract all peaks to plot in heatmap
# peaks <- peaks_bins %>% dplyr::filter(bin_ID %in% output_bins$anchor_I | bin_ID %in% output_bins$anchor_J)
# length(peaks$peak_ID)
# write.csv(peaks$peak_ID, "./output/Rshiny_PPR_input_peaks.csv")
# 
# # 1) Extract input_bins from NC genes
# input_bins <- extract_bins_from_features(features=NC_genes, feature_type="gene_name", bin_dictionary=promoters_bins)
# 
# # 2) See which output_bins are interacting with these input_bins (removes input bins)
# output_bins <- extract_interacting_bins(input_bins = input_bins, interactions = interactions)
# 
# # 3) See which features are in these output_bins
# output_features <- extract_features_from_bins(output_bins, peaks_bins, promoters_bins)
# peaks <- peaks_bins %>% dplyr::filter(bin_ID %in% output_bins)
# length(peaks$peak_ID)
# write.csv(peaks$peak_ID, "./output/Rshiny_NC_input_peaks.csv")
# 
# 
# 
# # for each interaction, how many interactions are between input_bins and how many are input_bin to output_bin
# investigate_interactions <- selected_interactions %>% 
#   mutate(anchor_I_in_input = anchor_I %in% input_bins$bin_ID) %>% 
#   mutate(anchor_J_in_input = anchor_J %in% input_bins$bin_ID) %>% 
#   mutate(Number_input_bins = rowSums(across(where(is.logical))))
# table(investigate_interactions$Number_input_bins)
